foreman/data_refinery_foreman/foreman/main.py

```diff
@@ -0,0 +1,260 @@
+    however if it takes longer to run than min_loop_time, then it will
+    be run less often than once per min_loop_time.
+    """
+    def decorator(function: Callable) -> Callable:
+        @wraps(function)
+        def wrapper(*args, **kwargs):
+            while(True):
+                start_time = timezone.now()
+
+                function(*args, **kwargs)
+
+                loop_time = timezone.now() - start_time
+                if loop_time < min_loop_time:
+                    remaining_time = MIN_LOOP_TIME - loop_time
+                    time.sleep(remaining_time.seconds)
+
+        return wrapper
+    return decorator
+
+
+@do_forever(MIN_LOOP_TIME)
+def retry_failed_downloader_jobs() -> None:
+    """Handle downloader jobs that were marked as a failure."""
+    failed_jobs = DownloaderJob.objects.filter(success=False, retried=False)
+    handle_downloader_jobs(failed_jobs)
+
+
+@do_forever(MIN_LOOP_TIME)
+def retry_hung_downloader_jobs() -> None:
+    """Retry downloader jobs that were started but never finished."""
+    minimum_start_time = timezone.now() - MAX_DOWNLOADER_RUN_TIME
+    hung_jobs = DownloaderJob.objects.filter(
+        success=None,
+        retried=False,
+        end_time=None,
+        start_time__lt=minimum_start_time
+    )
+
+    handle_downloader_jobs(hung_jobs)
+
+
+@do_forever(MIN_LOOP_TIME)
+def retry_lost_downloader_jobs() -> None:
+    """Retry downloader jobs that went too long without being started.
+
+    Idea: at some point this function could integrate with the spot
+    instances to determine if jobs are hanging due to a lack of
+    instances. A naive time-based implementation like this could end
+    up retrying every single queued job if there were a long period of
+    spot instance bid price being very high.
+    """
+    minimum_creation_time = timezone.now() - MAX_QUEUE_TIME
+    lost_jobs = DownloaderJob.objects.filter(
+        success=None,
+        retried=False,
+        start_time=None,
+        end_time=None,
+        created_at__lt=minimum_creation_time
+    )
+
+    handle_downloader_jobs(lost_jobs)
+
+
+@retry(stop_max_attempt_number=3)
+@transaction.atomic
+def requeue_processor_job(last_job: ProcessorJob) -> None:
+    """Queues a new downloader job and a Celery task for it.
+
+    The new downloader job will have num_retries one greater than
+    last_job.num_retries.
+    """
+    num_retries = last_job.num_retries + 1
+
+    new_job = ProcessorJob.create_job_and_relationships(num_retries=num_retries,
+                                                        batches=list(last_job.batches.all()),
+                                                        pipeline_applied=last_job.pipeline_applied)
+    processor_task = PROCESSOR_PIPELINE_LOOKUP[last_job.pipeline_applied]
+    app.send_task(processor_task, args=[new_job.id])
+
+    last_job.retried = True
+    last_job.success = False
+    last_job.save()
+
+
+def handle_processor_jobs(jobs: List[ProcessorJob]) -> None:
```

Comment:
I had the same thought several times. I've also thought about it in a more general sense too, as in do we really need to make any distinction between processor jobs and downloader jobs? However there are some small differences which would require many functions dealing with jobs to have handle them differently based on what they are. I think for the time being the overlap between the two is better than having an `if` statement in each function saying "is this job a processor or a downloader? ok, so then I set `pipeline_applied` instead of that `downloader_task`" or something.

If I really wanted to I'm sure I could get everything working in such a way that downloader jobs are just an instance of processor pipelines, but I think there might be some finagling that'd have to be done to make that work and I'm worried it'd be a lot harder to split the two in the future if there ever was a reason they did have to be two separate entities than it would be to merge them.